PerfettoCds/Pipeline/PerfettoSourceParser.cs

```diff
@@ -0,0 +1,145 @@
+                var shellDir = Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location);
+                var shellPath = Path.Combine(shellDir, PerfettoPluginConstants.TraceProcessorShellFileName);
+
+                // Start the Perfetto trace processor shell with the trace file
+                traceProc.OpenTraceProcessor(shellPath, filePath);
+
+                // Use this callback to receive events parsed and turn them in events with keys
+                // that get used by data cookers
+                void EventCallback(PerfettoSqlEvent ev)
+                {
+                    if (ev.GetType() == typeof(PerfettoSliceEvent))
+                    {
+                        // We get the timestamps used for displaying these events from the slice event
+                        if (traceStartTime == null)
+                        {
+                            traceStartTime = new Timestamp(((PerfettoSliceEvent)ev).Timestamp);
+                        }
+                        traceEndTime = new Timestamp(((PerfettoSliceEvent)ev).Timestamp);
+                    }
+
+                    PerfettoSqlEventKeyed newEvent = new PerfettoSqlEventKeyed(ev.GetEventKey(), ev);
+
+                    // Store the event
+                    var result = dataProcessor.ProcessDataElement(newEvent, this, cancellationToken);
+                }
+
+                // Perform queries for the events we need
+                List<PerfettoSqlEvent> eventsToQuery = new List<PerfettoSqlEvent> 
+                { 
+                    new PerfettoSliceEvent(), 
+                    new PerfettoArgEvent(), 
+                    new PerfettoThreadTrackEvent(), 
+                    new PerfettoThreadEvent(), 
+                    new PerfettoProcessEvent()
+                };
+
+                // Increment progress for each table queried.
+                double queryProgressIncrease = 99.0 / eventsToQuery.Count;
+
+                foreach (var query in eventsToQuery)
+                {
+                    logger.Info($"Performing Perfetto SQL query: {query.GetSqlQuery()}");
+
+                    // Run the query and process the events.
+                    traceProc.QueryTraceForEvents(query.GetSqlQuery(), query.GetEventKey(), EventCallback);
+
+                    IncreaseProgress(queryProgressIncrease);
+                }
+
+                // Done with the SQL trace processor
+                traceProc.CloseTraceConnection();
+
+                if (traceStartTime.HasValue)
+                {
+                    // Use DateTime.Now as the wall clock time. This doesn't matter for displaying events on a relative timescale
+                    // TODO Actual wall clock time needs to be gathered from SQL somehow
+                    this.dataSourceInfo = new DataSourceInfo(traceStartTime.Value.ToNanoseconds, traceEndTime.Value.ToNanoseconds, DateTime.Now.ToUniversalTime());
```

Comment:
The absolute time is incorrect for the trace and just is whenever the trace was opened. Replace "DateTime.Now.ToUniversalTime()" with the actual trace start time and verify event absolute timestamps in UTC match what other Perfetto tools provide

Reply:
I still haven't figured out how to get the absolute time the trace was taken from Perfetto. It is also unclear what the relative timestamps are relative to. I have a future task to figure out the timings. But for now the relative timestamps align correctly within the same trace.